src/plugins/point_cloud/PointCloud.cc

```diff
@@ -388,17 +389,24 @@ void PointCloudPrivate::PublishMarkers()
   auto minC = this->minColor;
   auto maxC = this->maxColor;
   auto floatRange = this->maxFloatV - this->minFloatV;
-  for (; iterX != iterX.End() &&
-         iterY != iterY.End() &&
-         iterZ != iterZ.End(); ++iterX, ++iterY, ++iterZ, ++ptIdx)
+  auto num_points =
+    this->pointCloudMsg.data().size() / this->pointCloudMsg.point_step();
+  if (num_points != this->floatVMsg.data().size())
+  {
+    gzwarn << "Float message and pointcloud are not of the same size,"
+      <<" visualization may not be accurate" << std::endl;
+  }
+  if (this->pointCloudMsg.data().size() % this->pointCloudMsg.point_step() != 0)
+  {
+    gzwarn << "Mal-formatted pointcloud" << std::endl;
+  }
+
+  for (; ptIdx < std::min<int>(this->floatVMsg.data().size(), num_points);
+    ++iterX, ++iterY, ++iterZ, ++ptIdx)
   {
     // Value from float vector, if available. Otherwise publish all data as
     // zeroes.
     float dataVal = 0.0;
-    if (this->floatVMsg.data().size() > ptIdx)
-    {
-      dataVal = this->floatVMsg.data(ptIdx);
```

Comment:
@arjo129 I understand why the `if` clause is no longer necessary, but why remove the assignment too?

Reply:
Oh no that was not intended :skull: will add some unit tests as well.

Reply:
My bad. This is wrong :skull: fixed.